src/newcharacter.cpp

```diff
@@ -1088,8 +1070,61 @@ int set_profession(WINDOW *w, player *u, int &points)
     ctxt.register_action("NEXT_TAB");
     ctxt.register_action("SORT");
     ctxt.register_action("HELP_KEYBINDINGS");
+    ctxt.register_action("FILTER");
+
+    bool recalc_profs = true;
+    int profs_length;
+    std::string filterstring = "";
+    std::vector<const profession *> sorted_profs;
 
     do {
+        if (recalc_profs) {
+            sorted_profs.clear();
+            for (profmap::const_iterator iter = profession::begin(); iter != profession::end(); ++iter) {
+                if ((g->scen->profsize() == 0 && (iter->second).has_flag("SCEN_ONLY") == false) ||
+                    g->scen->profquery(&(iter->second)) == true) {
+                    if (!filterstring.empty()) {
+                        if (!lcmatch(iter->second.gender_appropriate_name(u->male), filterstring)) {
+                            continue;
+                        }
+                    }
+                    sorted_profs.push_back(&(iter->second));
+                }
+            }
+            profs_length = sorted_profs.size();
+            if (profs_length == 0) {
+                popup(_("Nothing found.")); // another case of black box in tiles
+                filterstring.clear();
+                continue;
+            }
+
+            // Sort professions by points.
+            // profession_display_sort() keeps "unemployed" at the top.
+            profession_sorter.male = u->male;
+            std::sort(sorted_profs.begin(), sorted_profs.end(), profession_sorter);
```

Comment:
`std::stable_sort` would make this less chaotic (preserve order of professions) while not having a noticeable effect on performance.